From 235e306470be92bb759770ad076fc32740886faa Mon Sep 17 00:00:00 2001
From: Stefan Schneider <schneider@hrz.uni-marburg.de>
Date: Sat, 21 Nov 2015 13:58:29 +0100
Subject: [PATCH] The variable 'storage_engine' has been deleted since mysql
 5.7.5
 (https://dev.mysql.com/doc/refman/5.7/en/server-system-variables.html#sysvar_storage_engine).
 The variable 'default_storage_engine' could be used instead of
 'storage_engine'. It was introduced in mysql 5.5.3
 (https://dev.mysql.com/doc/refman/5.5/en/server-system-variables.html#sysvar_storage_engine)

---
 Services/Database/classes/class.ilDBInnoDB.php |  3 +--
 Services/Database/classes/class.ilDBMySQL.php  | 33 ++++++++++++++++++++++----
 2 files changed, 30 insertions(+), 6 deletions(-)

diff --git a/Services/Database/classes/class.ilDBInnoDB.php b/Services/Database/classes/class.ilDBInnoDB.php
index a482853..718066d 100755
--- a/Services/Database/classes/class.ilDBInnoDB.php
+++ b/Services/Database/classes/class.ilDBInnoDB.php
@@ -43,8 +43,7 @@ class ilDBInnoDB extends ilDBMySQL
 		{
 			$this->query("SET SESSION SQL_MODE = 'ONLY_FULL_GROUP_BY'");
 		}
-
-		$this->query("SET SESSION STORAGE_ENGINE = 'INNODB'");
+		$this->setStorageEngine('INNODB');
 	}
 
 	/**
diff --git a/Services/Database/classes/class.ilDBMySQL.php b/Services/Database/classes/class.ilDBMySQL.php
index 9aecf27..d8d932f 100755
--- a/Services/Database/classes/class.ilDBMySQL.php
+++ b/Services/Database/classes/class.ilDBMySQL.php
@@ -219,6 +219,14 @@ class ilDBMySQL extends ilDB
 	}
 	
 	/**
+	 * Set the storage engine 
+	 */
+	function setStorageEngine($a_storage_engine) {
+		$storage_engine_var = ($this->isMysql5_6OrHigher()) ? "DEFAULT_STORAGE_ENGINE" : "STORAGE_ENGINE";
+		$this->query("SET SESSION " . $storage_engine_var . " = '" . $a_storage_engine . "'");
+	}
+	
+	/**
 	* Get reserved words
 	*/
 	static function getReservedWords()
@@ -328,10 +336,9 @@ class ilDBMySQL extends ilDB
 		{
 			$this->query("SET SESSION SQL_MODE = 'ONLY_FULL_GROUP_BY'");
 		}
-
-		$this->query("SET SESSION STORAGE_ENGINE = 'MYISAM'");
+		$this->setStorageEngine('MYISAM');
 	}
-
+	
 	/**
 	* now()
 	*
@@ -415,7 +422,25 @@ class ilDBMySQL extends ilDB
 		
 		return false;
 	}
-
+	
+	/**
+	* check wether current MySQL server is version 5.6.x or higher
+	*
+	* NOTE: Two sourcecodes use this or a similar handling:
+	* - classes/class.ilDB.php
+	* - setup/classes/class.ilClient.php
+	*/
+	function isMysql5_6OrHigher()
+	{
+		$version = explode(".", $this->getDBVersion());
+		if ((int)$version[0] >= 5 ||
+			((int)$version[0] == 5 && (int)$version[1] >= 6))
+		{
+			return true;
+		}		
+		return false;
+	}
+	
 	/**
 	* Check query size
 	*/
-- 
2.4.9 (Apple Git-60)

